hazelcast/src/main/java/com/hazelcast/scheduledexecutor/impl/DistributedScheduledExecutorService.java

```diff
@@ -225,7 +225,9 @@ private void registerPartitionListener() {
                 new PartitionLostListener() {
                     @Override
                     public void partitionLost(PartitionLostEvent event) {
-                        for (ScheduledFutureProxy future : lossListeners) {
+                        // use toArray before iteration since it is done under mutex
```

Comment:
which mutex are you referring too? The mutex of the synchronousSet wrapper?

Reply:
the `mutex` of `SynchronizedCollection#toArray` inside java.util.Collections.

Reply:
Alternatively, we can synchronize on the collection itself. The javadoc for `java.util.Collections#synchronizedSet(java.util.Set<T>)` even says so:
```
     * It is imperative that the user manually synchronize on the returned
     * set when iterating over it:
     * <pre>
     *  Set s = Collections.synchronizedSet(new HashSet());
     *      ...
     *  synchronized (s) {
     *      Iterator i = s.iterator(); // Must be in the synchronized block
     *      while (i.hasNext())
     *          foo(i.next());
     *  }
     * </pre>
     * Failure to follow this advice may result in non-deterministic behavior.
```

Reply:
You will still run into a ConcurrentModificationException if you don't clone the set.

Reply:
The mutation methods synchronize on the same mutex (the collection itself).

Reply:
@mmedenjak it will work yes. My reasoning was to go out of critical section in a fixed time. Since notify methods that are called during iteration can be subject to change. It seemed creating a new array may not be an issue since i don't believe it is not on a hot path.